File: LoveLife/app/src/main/java/netUtil/baseObserver.java
package netUtil;

import android.util.Log;

import org.json.JSONException;
import org.json.JSONObject;

import io.reactivex.Observer;
import io.reactivex.disposables.Disposable;

/**
 * 封装Retrofit的Observer结合Rxjava
 * 当发送网络请求时，返回的结果会先从onNext方法执行
 * Created by Wuwei on 2018/6/26.
 */

public abstract class baseObserver<T> implements Observer<JSONObject> {

    private JSONObject mJSONObject;
    private String TAG = "baseObserver";

    @Override
    public void onSubscribe(Disposable d) {

    }

    @Override
    public void onNext(JSONObject response) {
        try {
            mJSONObject = new JSONObject(response.toString());
            switch (mJSONObject.getInt("status")){
                case 200:
                    //请求成功
                    onHandlerSeccess(mJSONObject);
                    break;
                case 403:
                    //Token错误
                    onHandlerFail(mJSONObject);
                    break;
                case 400:
                    //错误的请求
                    onHandlerFail(mJSONObject);
                    break;
            }
        } catch (JSONException e) {
            e.printStackTrace();
        }

    }

    //访问过程中的其他错误
    @Override
    public void onError(Throwable error) {
        Log.i(TAG, "onError: "+error.toString());
    }

    @Override
    public void onComplete() {

    }

    //暴漏在外部的请求成功时调用的方法
    public abstract void onHandlerSeccess(JSONObject response);

    //发送失败时调用的方法
    public abstract void onHandlerFail(JSONObject response);
}

File: LoveLife/app/src/main/java/netUtil/BaseObserver.kt
package netUtil

import android.util.Log
import com.google.gson.JsonObject
import io.reactivex.Observer
import io.reactivex.disposables.Disposable

/**
 *网路请求回调的类
 * Created by Wuwei on 2018/8/1.
 */
abstract class BaseObserver : Observer<JsonObject> {

    override fun onSubscribe(d: Disposable) {
    }

    override fun onComplete() {
    }

    override fun onNext(request : JsonObject) {
        Log.i("onNext","执行了"+request.toString())
        if (request.get("status").toString().equals("200")) onSuccess(request) else onFail(request)
    }

    override fun onError(e: Throwable) {
        Log.i("onError","BaseObserver执行了"+e.message)
    }


    abstract fun onSuccess(requeas : JsonObject)
    abstract fun onFail(requeas : JsonObject)

}